---
title: "tech_comparasion"
author: "Yu-Hsin Josch Hsieh"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
setwd("~/Ludwig_lab/scmtMMB/PBMC/code/")
```

### Load packages

```{r packages, message=FALSE, warning=FALSE}
library(Matrix)
library(dplyr)
library(tidyr)
library(SummarizedExperiment)
library(ggplot2)
library(viridis)
library(gridExtra)
require(scales)
'%ni%' <- Negate('%in%')
source("../../global_func/quantifyMMB.R")
source("../../global_func/variant_calling.R")
source("../../global_func/mtGene_color.R")

```

```{r mito data loading, message=FALSE, warning=FALSE}
# Load the mitochondrial genome data from multiome
ATAC_meta <- read.csv("../PBMC_large_data_files/input/Healthy_Y05_mtMultiome/Seurat/Hped_31697_mtMultiome_pred.celltype.l2.csv", row.names = "X")
ATAC_mgatk <- readRDS("../PBMC_large_data_files/input/Healthy_Y05_mtMultiome/mgatk/mgatk.rds")
ATAC_mmat <- call_mutations_mgatk(ATAC_mgatk)
ATAC_cov_per_pos_cell <- assays(ATAC_mgatk)[['coverage']]
ATAC_var <- as.data.frame(rowData(ATAC_mmat)) %>% subset(n_cells_conf_detected >= 1 &
                          strand_correlation >= 0.65 & #vmr >= 0.01 &
                          variant %ni% c("301A>C", "302A>C", "310T>C", "316G>C")) %>% pull(variant)
ATAC_mmat <- ATAC_mmat[ATAC_var, rownames(ATAC_meta)]
ATAC_mmat <- ATAC_mmat[,colData(ATAC_mmat)$depth>5]

RNA_mgatk <- readRDS("../PBMC_large_data_files/input/Healthy_Y05_mtMultiome/mgatk/rna.mgatk.rds")[,colnames(ATAC_mmat)]
RNA_mmat <- call_mutations_mgatk(RNA_mgatk)
RNA_cov_per_pos_cell <- assays(RNA_mgatk)[['coverage']]

as.data.frame(rowData(RNA_mmat)) %>% 
  ggplot(aes(mean_coverage, mean)) + geom_point() + 
  geom_vline(xintercept = 10, linetype = "longdash") +
  geom_hline(yintercept = 0.05, linetype = "longdash")+
  theme_classic() 

data_frame(x = assays(RNA_mmat)$coverage["2617A>G",]) %>%
    ggplot(., aes(x)) + 
    geom_bar(col="black", fill = "grey") + theme_classic()


RNA_var <- as.data.frame(rowData(RNA_mmat)) %>% 
  mutate(n_cells_conf_detected = rowSums(assays(RNA_mmat)$coverage > 10)) %>% 
  filter(n_cells_conf_detected >= 10 & variant %ni% c("301A>C", "302A>C", "310T>C", "316G>C") & mean > 0 & mean_coverage >= 5) %>%
  pull(variant)
RNA_mmat_filtered <- RNA_mmat[RNA_var,]

as.data.frame(rowData(RNA_mmat_filtered)) %>% 
  ggplot(aes(mean_coverage, mean)) + geom_point() + geom_vline(xintercept = 5, linetype = "longdash") +
  theme_classic() 

# mtscATAC
mtscATAC <- readRDS("../PBMC_large_data_files/output/01_seurat_H05.rds")
mtscATAC_mgatk <- readRDS("../PBMC_large_data_files/input/Healthy_Y05/mgatk/mgatk.rds")[,colnames(mtscATAC)]
mtscATAC_mmat <- call_mutations_mgatk(mtscATAC_mgatk)
mtscATAC_var <- as.data.frame(rowData(mtscATAC_mmat)) %>% subset(n_cells_conf_detected >= 1 &
                          strand_correlation >= 0.65 & #vmr >= 0.01 &
                          variant %ni% c("301A>C", "302A>C", "310T>C", "316G>C")) %>% pull(variant)
mtscATAC_mmat <- mtscATAC_mmat[mtscATAC_var,]

mtscATAC_cov_per_pos_cell <- readRDS("../PBMC_large_data_files/output/01_mtDNA_pos_coverage_H05.rds")[,colnames(mtscATAC_mmat)]
mtscATAC_var <- as.data.frame(rowData(mtscATAC_mmat)) %>% pull(variant)

Venn <- BioVenn::draw.venn(
  mtscATAC_var, ATAC_var, RNA_var,
  title = NULL, subtitle = NULL,
  xtitle = NULL, ytitle = NULL, ztitle = NULL, 
  nr_c = "black", x_c = "blue", y_c = "deepskyblue", z_c = "red",
  bg_c = "white", output = "pdf", filename = "../plot/06_mtVar_venn.pdf", width = 800, height =800
)

```

```{r mito data coverage, message=FALSE, warning=FALSE}

pull_coverage <- function(SE, cells, resolution = 5){
  zoo::rollmean(rowMeans(assays(SE)[['coverage']]), resolution)
}

cov_df <- data.frame(
  pos = zoo::rollmean(1:16569, 5),
  RNA = pull_coverage(RNA_mgatk),
  ATAC = pull_coverage(ATAC_mgatk),
  mtscATAC = pull_coverage(mtscATAC_mgatk))

# visualize coverage
p1 <- cov_df %>% reshape2::melt(id.vars = "pos") %>% # dplyr::filter(value > 5) %>%
  ggplot() +
  geom_line(aes(x = pos, y = value, color = variable)) +
  scale_y_continuous(limits = c(-20, 52), expand = c(0,0)) +
  scale_x_continuous(breaks = seq(0,16596, by = 10000), expand = c(0,0)) + 
  scale_color_manual(values = c("red", "deepskyblue", "blue")) +
  theme_classic() +
  labs(x = "Position on mtDNA chromosome", y = "Roll mean coverage / cell")  + geom_rect(data = mitochondrial_data, 
            aes(xmin = Starting, xmax = Ending, ymin = -3, ymax = 0, fill = color), alpha = 1) +
  scale_fill_manual(values = mtGene_colors) +
  coord_polar() + 
  guides(fill=guide_legend(ncol=2))+
  theme_minimal()

p2 <- p1 + Seurat::NoLegend()
leg <- ggpubr::get_legend(p1)
p3 <- ggpubr::as_ggplot(leg)

ggsave("../plot/06_mtDNA_mtRNA_cov.pdf", p2, width = 5, height = 5)
ggsave("../plot/06_mtDNA_mtRNA_cov_leg.pdf", p3, width = 5, height = 5)
```

```{r mito variant, message=FALSE, warning=FALSE}

mean.df <- full_join(as.data.frame(rowData(ATAC_mmat)[,c("variant","mean", "mean_coverage")]),
as.data.frame(rowData(RNA_mmat)[,c("variant","mean", "mean_coverage")]),
by = "variant", suffix = c(".ATAC", ".RNA")) %>% replace(is.na(.), 0) %>% 
  full_join(., as.data.frame(rowData(mtscATAC_mmat)[,c("variant","mean", "mean_coverage")]),
by = "variant", suffix = c("", ".mtscATAC")) %>% replace(is.na(.), 0)


```

```{r stats, message=FALSE, warning=FALSE}
vaf.df <- data.frame(RNA = colSums(assays(RNA_mmat_filtered)$allele_frequency >0),
           ATAC = colSums(assays(ATAC_mmat_filtered)$allele_frequency>0)) 
vaf.df %>% ggplot(aes(x = RNA, y = ATAC)) +geom_point()
```

```{r mito variant1 , message=FALSE, warning=FALSE, fig.width= 5, fig.height=5}
ggplot(mean.df, aes(mean.ATAC, mean.RNA)) + geom_point()
ggplot(mean.df, aes(mean_coverage.RNA, mean.RNA)) + geom_point() + theme_classic()
ggplot(mean.df, aes(mean_coverage.ATAC, mean.ATAC)) + geom_point() + theme_classic()

ggplot(mean.df, aes(mean.ATAC, mean)) + geom_point() + xlim(0, 0.005) + ylim(0, 0.005) 

```

```{r mito variant 2, message=FALSE, warning=FALSE}
mean.df %>% filter(mean.RNA > 0.1 & mean.ATAC < 0.9) %>% arrange(desc(mean.RNA))
assays(RNA_mmat)$allele_frequency["2617A>G",][assays(RNA_mmat)$coverage["2617A>G",] > 4]
assays(RNA_mmat)$coverage["2617A>G",][assays(RNA_mmat)$coverage["2617A>G",] > 4]
assays(RNA_mmat)$allele_frequency["2617A>T",][assays(RNA_mmat)$coverage["2617A>T",] > 4]
assays(RNA_mmat)$coverage["2617A>T",][assays(RNA_mmat)$coverage["2617A>T",] > 4]
assays(RNA_mmat)$allele_frequency["3590T>A",][assays(RNA_mmat)$coverage["3590T>A",] > 4]
assays(RNA_mmat)$coverage["3590T>A",][assays(RNA_mmat)$coverage["3590T>A",] > 4]


```

